File: app/src/main/java/dev/ragnarok/fenrir/api/interfaces/INewsfeedApi.java
package dev.ragnarok.fenrir.api.interfaces;

import androidx.annotation.CheckResult;

import java.util.Collection;

import dev.ragnarok.fenrir.api.model.Items;
import dev.ragnarok.fenrir.api.model.VkApiFeedList;
import dev.ragnarok.fenrir.api.model.response.NewsfeedCommentsResponse;
import dev.ragnarok.fenrir.api.model.response.NewsfeedResponse;
import dev.ragnarok.fenrir.api.model.response.NewsfeedSearchResponse;
import io.reactivex.rxjava3.core.Single;

public interface INewsfeedApi {

    @CheckResult
    Single<Items<VkApiFeedList>> getLists(Collection<Integer> listIds);

    @CheckResult
    Single<NewsfeedSearchResponse> search(String query, Boolean extended, Integer count,
                                          Double latitude, Double longitude, Long startTime,
                                          Long endTime, String startFrom, String fields);

    @CheckResult
    Single<Integer> saveList(String title, Collection<Integer> listIds);

    @CheckResult
    Single<Integer> deleteList(Integer list_id);

    @CheckResult
    Single<NewsfeedCommentsResponse> getComments(Integer count, String filters, String reposts,
                                                 Long startTime, Long endTime, Integer lastCommentsCount,
                                                 String startFrom, String fields);

    @CheckResult
    Single<NewsfeedCommentsResponse> getMentions(Integer owner_id, Integer count, Integer offset, Long startTime, Long endTime);

    @CheckResult
    Single<NewsfeedResponse> get(String filters, Boolean returnBanned, Long startTime, Long endTime,
                                 Integer maxPhotoCount, String sourceIds, String startFrom, Integer count, String fields);

    @CheckResult
    Single<NewsfeedResponse> getRecommended(Long startTime, Long endTime,
                                            Integer maxPhotoCount, String startFrom, Integer count, String fields);

    @CheckResult
    Single<NewsfeedResponse> getFeedLikes(Integer maxPhotoCount, String startFrom, Integer count, String fields);

    @CheckResult
    Single<Integer> addBan(Collection<Integer> listIds);

    @CheckResult
    Single<Integer> ignoreItem(String type, Integer owner_id, Integer item_id);

}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/api/interfaces/INewsfeedApi.kt
package dev.ragnarok.fenrir.api.interfaces

import androidx.annotation.CheckResult
import dev.ragnarok.fenrir.api.model.Items
import dev.ragnarok.fenrir.api.model.VKApiFeedList
import dev.ragnarok.fenrir.api.model.response.NewsfeedBanResponse
import dev.ragnarok.fenrir.api.model.response.NewsfeedCommentsResponse
import dev.ragnarok.fenrir.api.model.response.NewsfeedResponse
import dev.ragnarok.fenrir.api.model.response.NewsfeedSearchResponse
import io.reactivex.rxjava3.core.Single

interface INewsfeedApi {
    @CheckResult
    fun getLists(listIds: Collection<Int>?): Single<Items<VKApiFeedList>>

    @CheckResult
    fun search(
        query: String?, extended: Boolean?, count: Int?,
        latitude: Double?, longitude: Double?, startTime: Long?,
        endTime: Long?, startFrom: String?, fields: String?
    ): Single<NewsfeedSearchResponse>

    @CheckResult
    fun saveList(title: String?, listIds: Collection<Long>?): Single<Int>

    @CheckResult
    fun deleteList(list_id: Int?): Single<Int>

    @CheckResult
    fun getComments(
        count: Int?, filters: String?, reposts: String?,
        startTime: Long?, endTime: Long?, lastCommentsCount: Int?,
        startFrom: String?, fields: String?
    ): Single<NewsfeedCommentsResponse>

    @CheckResult
    fun getMentions(
        owner_id: Long?,
        count: Int?,
        offset: Int?,
        startTime: Long?,
        endTime: Long?
    ): Single<NewsfeedCommentsResponse>

    @CheckResult
    operator fun get(
        filters: String?, returnBanned: Boolean?, startTime: Long?, endTime: Long?,
        maxPhotoCount: Int?, sourceIds: String?, startFrom: String?, count: Int?, fields: String?
    ): Single<NewsfeedResponse>

    @CheckResult
    fun getTop(
        filters: String?, returnBanned: Boolean?, startTime: Long?,
        endTime: Long?, maxPhotoCount: Int?, sourceIds: String?,
        startFrom: String?, count: Int?, fields: String?
    ): Single<NewsfeedResponse>

    @CheckResult
    fun getRecommended(
        startTime: Long?, endTime: Long?,
        maxPhotoCount: Int?, startFrom: String?, count: Int?, fields: String?
    ): Single<NewsfeedResponse>

    @CheckResult
    fun getFeedLikes(
        maxPhotoCount: Int?,
        startFrom: String?,
        count: Int?,
        fields: String?
    ): Single<NewsfeedResponse>

    @CheckResult
    fun addBan(listIds: Collection<Long>): Single<Int>

    @CheckResult
    fun ignoreItem(type: String?, owner_id: Long?, item_id: Int?): Single<Int>

    @CheckResult
    fun deleteBan(listIds: Collection<Long>): Single<Int>

    @CheckResult
    fun getBanned(): Single<NewsfeedBanResponse>
}